code/game/machinery/_machines_base/machinery_damage.dm

```diff
@@ -0,0 +1,79 @@
+/obj/machinery/proc/take_damage(amount, damtype, silent)
+	//Let's not bother initializing all the components for nothing
+	if(amount <= 0)
+		return
+	if(damtype != BRUTE && damtype != BURN && damtype != ELECTROCUTE)
+		return
+	if(!silent)
+		var/hitsound = 'sound/weapons/smash.ogg'
+		if(damtype == ELECTROCUTE)
+			hitsound = "sparks"
+		else if(damtype == BURN)
+			hitsound = 'sound/items/Welder.ogg'
+		playsound(src, hitsound, 10, 1)
+		
+	world << "[src] is taking [amount] of [damtype] damage"
+	// Shielding components (armor/fuses) take first hit
+	var/list/shielding = get_all_components_of_type(/obj/item/stock_parts/shielding)
+	world << "Shields: [json_encode(shielding)]"
+	for(var/obj/item/stock_parts/shielding/soak in shielding)
+		if(damtype in soak.protection_types)
+			amount -= soak.take_damage(amount, damtype)
+	if(amount <= 0)
+		return
+
+	// If some damage got past, next it's generic (non-circuitboard) components
+	world << "Getting meat"
+	var/obj/item/stock_parts/victim = get_damageable_component()
+	while(amount > 0 && victim)
+		world << "Got [victim]"
+		amount -= victim.take_damage(amount, damtype)
+		victim = get_damageable_component()
+	if(amount <= 0)
+		return
+
+	// And lastly hit the circuitboard
+	victim = get_component_of_type(/obj/item/stock_parts/circuitboard)
+	if(victim)
+		world << "Damaging circuitboard"
+		victim.take_damage(amount, damtype)
+	
+/obj/machinery/proc/get_damageable_component()
+	var/list/victims = component_parts.Copy()
+	victims = shuffle(victims)
+	for(var/obj/item/stock_parts/component in victims)
+		// Circuitboards are handled separately
+		if(istype(component, /obj/item/stock_parts/circuitboard))
+			continue
+		// Don't damage what can't be repaired
+		if(!(component.part_flags & PART_FLAG_HAND_REMOVE))
+			continue
+		if(component.is_functional())
+			return component
+	for(var/path in uncreated_component_parts)
+		if(uncreated_component_parts[path])
+			return force_init_component(path)
+
+/obj/machinery/proc/on_component_failure(var/obj/item/stock_parts/component)
+	RefreshParts()
+	power_change()
+
+/obj/machinery/emp_act(severity)
+	if(use_power && stat == 0)
+		new /obj/effect/temp_visual/emp_burst(loc)
+		use_power_oneoff(7500/severity)
+		take_damage(100/severity, ELECTROCUTE)
+	..()
+
+/obj/machinery/ex_act(severity)
+	take_damage(100/severity, BRUTE, TRUE)
+
+/obj/machinery/bullet_act(obj/item/projectile/P, def_zone)
+	. = ..()
+	take_damage(P.damage, P.damage_type)
+
+/obj/machinery/attackby(obj/item/W, mob/user)
+	..()
+	if(!(W.item_flags & ITEM_FLAG_NO_BLUDGEON) && W.force > 5)
+		user.setClickCooldown(W.attack_cooldown + W.w_class)
+		take_damage(W.force, W.damtype)
```

Comment:
well usually things only call ..() in attackby if they didn't do any special interactions, since it'll end at 'Urist hits thing with rock!' handling.
Will check if it gets in the way anywhere, can slap intent check.

Reply:
Just think it will be annoying when you mess up some interaction and it falls through to here. When you get that red message, at least you don't have to fix the machine afterwards.

Reply:
done